Blog: An Introduction to Observability for LLM-based applications using OpenTelemetry (#4455)
open-telemetry/opentelemetry.io: content/en/blog/2024/llm-observability.md

---

> title: Strategy for monitoring LLM Application with OpenTelemetry

Reviewer: ```suggestion
title: Observability for LLM-based applications using OpenTelemetry: An example
```

Or similar. Best to steer away from "monitor", as o11y is a wider concept.

linkTitle: LLM Observability
date: 2024-04-09
author: '[Ishan Jain](https://github.com/ishanjainn) (Grafana)'
# prettier-ignore
cSpell:ignore: associated chatbots Ishan ishan_jainn ishanjainn Jain llm timeframe
---

In the tech world, Large Language Models (LLMs), like the brains behind chatbots
or code generators, are getting big. But as they grow, making sure they keep

> running smoothly is a big deal.

Reviewer: Not sure if I get this sentence:

> like the brains behind chatbots or code generators, are getting big

Could you rephrase to state what's critical about LLMs and their infrastructure needs?

In the following example, we'll use [Prometheus](https://prometheus.io/) and
[Grafana Tempo](https://grafana.com/oss/tempo/) as the target backend for

> metrics and traces generated by an Auto-instrumentation LLM monitoring library

Reviewer: ```suggestion
metrics and traces generated by an auto-instrumentation LLM monitoring library
```

[OpenLIT](https://github.com/openlit/openlit). We will use

> [Grafana](https://grafana.com/oss/grafana/) as the tool to visualize the LLM

Reviewer: See my last comment about not using a specific backend.

monitoring data.

## Why Observability Matters for LLM Applications

Monitoring LLM applications is crucial for several reasons. For starters, it's
vital to keep track of how often LLMs are being used, especially in RAG-based
applications. This monitoring helps you understand the demand and usage
patterns. Secondly, knowing how long it takes for LLMs to respond to requests
and the nature of each interaction (prompt and response) can significantly
impact user experience and operational efficiency. Since using these models
incurs costs per request, tracking usage helps in managing expenses. Moreover,
making too many requests in a short timeframe can lead to rate limiting,
temporarily blocking further requests. By keeping a close eye on these aspects,
you can not only save costs but also avoid hitting request limits, ensuring your

> LLM applications perform optimally.

Reviewer: This part is very interesting. I wish you'd develop it more, for example using lists. What are the most important things to pay attention to and what are the related OTel signals?

traces to Grafana Tempo, follow these steps to visualize them in Grafana. You
can use any tool of your choice to visualize this data:

### Add Prometheus as a data source

1. In Grafana, navigate to **Connections** > **Data Sources**.
2. Click **Add data source** and select **Prometheus**.
3. In the settings, enter your Prometheus URL, for example,
   `http://<your_prometheus_host>`, along with any other necessary details.
4. Select **Save & Test**.

### Add Tempo as a data source

1. In Grafana, navigate to **Connections** > **Data Sources**.
2. Click **Add data source** and select **Tempo**.
3. In the settings, enter your Tempo URL, for example,
   `http://<your_tempo_host>`, along with any other necessary details.
4. Select **Save & Test**.

### Add the dashboard

To make things easy, you can use the dashboard the OpenLIT team made. Just grab
the JSON from
[here](https://docs.openlit.io/latest/connections/prometheus-tempo#dashboard).

This guide showed you how to use OpenTelemetry, Prometheus, Tempo, and Grafana
to monitor your LLM Applications.

If you have any questions, reach out on my GitHub
[@ishanjainn](https://github.com/ishanjainn) or Twitter

> [@ishan_jainn](https://twitter.com/ishan_jainn).

Reviewer: I think the whole backend or visualization part is unnecessary. OTel is vendor and backend agnostic. What's more, I don't think visualizing sample data in this case helps in supporting the content. Consider removing this part or rewriting to explain how the data can be consumed by any backend.

---

> title: Observability for LLM-based applications using OpenTelemetry

Reviewer 2: ```suggestion
title: An Introduction to Observability for LLM-based applications using OpenTelemetry
```

Reviewer 2: Rewording because this is a fairly deep topic, but the post contents are introductory in nature.

linkTitle: LLM Observability
date: 2024-04-09
author: '[Ishan Jain](https://github.com/ishanjainn) (Grafana)'
# prettier-ignore
cSpell:ignore: associated chatbots Ishan ishan_jainn ishanjainn Jain llm timeframe
---

As Large Language Models (LLMs) rapidly transform the technology landscape by
driving advanced chatbots and innovative code generators, their expanding role
underscores the necessity for robust observability mechanisms. Ensuring the
optimal performance and reliability of LLM-based applications is not just

> beneficial but essential in today's tech-driven environment.

Reviewer 2: This reads like it was just output by chatgpt with default settings. Could you reword this to not use such flowery language?

In the following example, we'll use [Prometheus](https://prometheus.io/) and

> [Grafana Tempo](https://grafana.com/oss/tempo/) as the target backend for

Reviewer 2: Can you switch this to Jaeger? While temp is OSS, this is more of a vendor blog post than I think we'd prefer.

metrics and traces generated by an auto-instrumentation LLM monitoring library
[OpenLIT](https://github.com/openlit/openlit). We will use
[Grafana](https://grafana.com/oss/grafana/) as the tool to visualize the LLM
monitoring data. You can choose any backend of your choice to store OTel metrics
and traces.

## Why Observability Matters for LLM Applications

> Monitoring LLM applications is crucial for several reasons.

Reviewer 2: Note: monitoring != observability

1. It's vital to keep track of how often LLMs are being used, especially in
   RAG-based applications. This monitoring helps you understand the demand and

>    usage patterns.

Reviewer 2: >  especially in RAG-based applications

Why? In my experience building LLM-based products there's nothing unique about RAG here.

Moreoever, what usage patterns are you referring to?

2. Understanding how long it takes for LLMs to respond to requests and the
   nature of each interaction (prompt and response) can significantly impact
   user experience and operational efficiency. Since using these models incurs

>    costs per request, tracking usage helps in managing expenses.

Reviewer 2: I think you can shorten some of this text to just say that latency is important for these applications since it can vary so wildly based on inputs.

3. Making too many requests in a short timeframe can lead to rate limiting,

>    temporarily blocking further requests.

Reviewer 2: This isn't particularly unique for LLM-based applications, and when you self-host a fine-tuned model it's not really a concern either.

By keeping a close eye on these aspects, you can not only save costs but also
avoid hitting request limits, ensuring your LLM applications perform optimally.

> ## What are the signal that you should be looking at?

Reviewer 2: ```suggestion
## What are the signals that you should be looking at?
```

Integrating Large Language Models (LLMs) into applications represents a pivotal
shift from traditional machine learning (ML) practices, particularly because
these LLMs are accessed via API calls rather than being hosted locally or run
in-house. This external integration necessitates a nuanced approach to
observability, where understanding both the sequence of events (through traces)
and the aggregated data (through metrics) becomes crucial for optimizing

> performance and managing costs. Here's the critical signals to monitor:

Reviewer 2: Similarly to the first paragraph, this reads as chatgpt default settings.

### Traces

Tracing is vital for capturing the flow of requests through your system and
understanding the interactions with hosted LLM services. Here's what you need to

> focus on:

Reviewer 2: Can you elaborate on why tracing matters here? "flow of requests" doesn't say much. Above you mention how RAG is important, can you connect that here?

- **Request Metadata**: This is even more critical in the context of LLMs than
  traditional ML, given the variety of parameters (like temperature and prompt
  details) that can drastically affect both the response quality and the cost.

>   Specific aspects to monitor are:

Reviewer 2: > This is even more critical in the context of LLMs than traditional ML

I don't think this is true at all. Many different kinds of models (and not just hosted models like GPT) have plenty of request parameters that affect outputs.

>   - **Temperature**: Indicates the level of creativity or randomness desired

Reviewer 2: top_p sampling is equally as important to capture.

    from the model's outputs. Varying this parameter can significantly impact
    the nature of the generated content.
  - **Model Name or Version**: Essential for tracking over time, as updates to
    the LLM might affect performance or response characteristics.
  - **Prompt Details**: The exact inputs sent to the LLM, which, unlike in-house
    ML models where inputs might be more controlled and homogeneous, can vary

>     wildly and affect output complexity and cost implications.

Reviewer 2: Note that in the LLM Working Group, we're explicitly not capturing these details on spans because many backend systems can struggle with those often large payloads.

- **Response Metadata**: Given the API-based interaction with LLMs, tracking the
  specifics of the response is key for cost management and quality assessment:
  - **Tokens**: Directly impacts cost and is a measure of response length and
    complexity.
  - **Cost**: Critical for budgeting, as API-based costs can scale with the

>     number of requests and the complexity of each request.

Reviewer 2: Why capture cost here when it's captured as a metric?

Cost is also per-request in a tracing model, but here you describe it as an aggregate.

  - **Response Details**: Similar to the prompt details but from the response
    perspective, providing insights into the model's output characteristics and

>     potential areas of inefficiency or unexpected cost.

Reviewer 2: Similar to above statement, these are being designed to be captured as events, not attributes on a span.
